Add forked_daapd integration

uvjustin opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Added a forked_daapd component (and pypi library) to support forked_daapd as a media_device. This integration is primarily used to control the forked-daapd outputs (zones) with limited playback control (play/pause, previous/next track) and media info. Playlist manipulation is not supported.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml
media_player:
  - platform: forked_daapd
    host: 192.168.1.60
```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->
A few notes:
- Just need to provide this component the IP and optional port of the forked-daapd server/API. The output list and the websocket connection port will be automatically loaded from the API.
- Uses asyncio with async/await syntax
- Uses callbacks for updates, not polling
- Maintains a websocket connection to the forked-daapd server to listen for updates

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]: https://github.com/home-assistant/home-assistant.io/pull/12163

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

uvjustin committed:
New forked_daapd component

springstan commented:
Added `docs-missing` so we do not forget to update the documentation before merging :)

uvjustin commented:
@springstan I'm new to developing...Can you help me answer two questions?
1) How do I get codecov/patch to pass? I understand that it's related to testing but I'm not sure how to implement real testing on this library as the ultimate functionality depends on both the PYPI library and a running forked-daapd server. I've noticed that some other integrations have very simple tests or don't have tests defined at all and those seem to pass.
2) Do I need to write my own documentation and create a separate PR for that?

springstan commented:
Sure thing @uvjustin!

1. First of all code coverage means how much of the code is covered by the tests that are written for a particular set of files. You are completely right that you cannot rely on the library and a running server during testing that is why [mocking](https://realpython.com/python-mock-library/) is used. With mocking you can imitate the behavior of the library and the server. The tests can be rather simple if they cover the code that is totally fine.

2. Yes, indeed in the [documentation repo](https://github.com/home-assistant/home-assistant.io).

natedreger commented:
this will be great for me! Wish I could help.

escoand commented:
Tested your integration and it worked out quite well, but I've some comments:

- when the daap is off when starting HA the entity is not created and never will
- are the zones dynamical, will they be created when detected?
- as the config doesn't know a name attribute it would be nice you could add a [unique id](https://developers.home-assistant.io/docs/entity_registry_index/#unique-id-requirements) so we could rename the entity on the UI
- automatic detection via [zeroconf](https://developers.home-assistant.io/docs/en/next/creating_integration_manifest.html#zeroconf) would be great, `_dacp._tcp.` should be the correct type, maybe we have to filter out forked-daapd

uvjustin commented:
@natedreger @escoand Great to see there is some interest in this. I wasn't sure if anyone would use it. I've added some functionality locally like getting it to work with TTS (broadcast announcements throughout the house) and controlling music that's piped in via librespot-java (great library that is very actively maintained by @devgianlu https://github.com/librespot-org/librespot-java/).
Will try to eventually add the features @escoand mentioned above. As per the zones, they are loaded from the forked-daapd server config.

escoand commented:
Yes, I know they are loaded, but only at start or also later again?

uvjustin commented:
Oh currently just at the start. I can make changes so it will add them dynamically.

escoand commented:
There is one more zeroconf advertisement:
![Screenshot_20200324-124433](https://user-images.githubusercontent.com/428567/77422294-7fea3480-6dcd-11ea-8b21-11c01bf66c30.png)

escoand commented:
I realized the integration need quite long to recognize an activated or deactivated output. Seems like the websocket push notifications do not work.

MartinHjelmare commented on homeassistant/components/forked_daapd/media_player.py:
```diff
@@ -0,0 +1,476 @@
+DOMAIN = "forked_daapd"
+DEFAULT_PORT = 3689
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
```
We no longer accept additions or changes to the yaml config of integrations. See: https://github.com/home-assistant/architecture/blob/master/adr/0010-integration-configuration.md

This integration needs to be refactored to use a config flow and config entries.

https://developers.home-assistant.io/docs/config_entries_config_flow_handler

uvjustin committed:
Bunch of changes

Add config flow and zeroconf
Add zones on callback when added by server
Add password auth
Add async_play_media for TTS
Add media_image_url
Add support for pipe control/input from librespot-java
Improve update callbacks

uvjustin commented:
Pushed a bunch of changes and added config flow/config entry support

MartinHjelmare commented on tests/components/forked_daapd/test_media_player.py:
```diff
@@ -0,0 +1,559 @@
+
+    def async_add_entities(list_of_entities, update_on_add):
+        for entity in list_of_entities:
+            print(entity)
```
Please remove all use of print in the tests.

MartinHjelmare commented on tests/components/forked_daapd/test_media_player.py:
```diff
@@ -0,0 +1,559 @@
+        "homeassistant.components.forked_daapd.media_player.ForkedDaapdAPI",
+        autospec=True,
+    ):
+        device = ForkedDaapdDevice(
```
Don't create the entity directly. Set up the integration and platform and let it add an entity. Then assert the state of the entity via the core state machine.

See eg:
https://github.com/home-assistant/core/blob/4484afc0d6756f2df140925285226cd2d1eb1e13/tests/components/heos/test_media_player.py#L66-L100

MartinHjelmare reviewed (commented):
The config flow requires 100% test coverage. For the other modules tests are not required but are appreciated.

uvjustin commented on homeassistant/components/forked_daapd/__init__.py:
```diff
@@ -0,0 +1,20 @@
+
+async def async_setup_entry(hass, entry):
+    """Set up forked-daapd from a config entry by forwarding to platform."""
+    hass.async_add_job(hass.config_entries.async_forward_entry_setup(entry, MP_DOMAIN))
```
Thanks, I will make the change. I got async_add_job from here https://developers.home-assistant.io/docs/config_entries_index/ , can we change the documentation?

MartinHjelmare replied:
Yes, if you have time, please make a PR to the dev docs. :+1:

uvjustin commented on homeassistant/components/forked_daapd/media_player.py:
```diff
@@ -0,0 +1,695 @@
+    async def async_toggle(self):
+        """Toggle the power on the zone.
+
+        Default media player component method counts idle as off - we consider idle to be on but just not playing.
```
Is there a way to get this checked automatically? I thought black would take care of it.

MartinHjelmare replied:
Black handles everything except long string length, I think. We need to take care of that ourselves.

uvjustin commented on homeassistant/components/forked_daapd/media_player.py:
```diff
@@ -0,0 +1,695 @@
+        if media_type == MEDIA_TYPE_MUSIC:
+            saved_state = self.state  # save play state
+            await self.async_turn_off()  # pauses and saves output states
+            await self._set_tts_volumes()
```
The use case for async_play_media is just for playing a TTS media url. The method pauses the playback (if playing) and loads in the TTS media, then resumes normal playback (if playing).

MartinHjelmare replied:
@balloob is this ok?

MartinHjelmare replied:
@balloob do you see any problems with this? I haven't seen this before so I'm not sure about it.

uvjustin commented on homeassistant/components/forked_daapd/media_player.py:
```diff
@@ -0,0 +1,695 @@
+            _LOGGER.debug("Could not get track info.")
+            self._data.track_info = defaultdict(str)
+
+    update_callback = partialmethod(_update)
```
It's to allow the callback sent to the pypi library to reference self.

MartinHjelmare replied:
Sorry, I don't understand. It does not work to give `self._update` as callback?

uvjustin replied:
You are right, I should be able to pass in self._update. I meant I used partialmethod instead of partial because I wanted to pass self. I think I was using that because I was freezing something else in the call that I took out subsequently.

uvjustin commented:
Thanks for the comments, will work on putting them through

uvjustin commented on homeassistant/components/forked_daapd/media_player.py:
```diff
@@ -0,0 +1,695 @@
+        """Return whether the zone is available."""
+        return self._available
+
+    @available.setter
+    def available(self, val):
+        """Setter for property."""
+        self._available = val
+
```
Ok. Should I just remove the getter here too and rename the _available variable to available to allow direct access?

MartinHjelmare replied:
No we need to keep the `available` property since it's part of the base entity class. Make a private attribute `self._available` that we return for that property and that we can change as needed. The change should come from inside the entity. If we need to do it from outside the entity we use our dispatch helper signals for that.

uvjustin committed:
Refactor as per code review suggestions

Move config_flow connection testing to pypi library (v0.1.4)
Remove use of ForkedDaapdData class
Decouple Master and Zone data and functions
Add updater class to manage websocket and entity updates

uvjustin commented:
Documentation PR https://github.com/home-assistant/home-assistant.io/pull/12163

MartinHjelmare commented on homeassistant/components/forked_daapd/media_player.py:
```diff
@@ -0,0 +1,669 @@
+                self.hass, SIGNAL_UPDATE_MASTER, self._update_callback
+            )
+        )
+        self._updater = ForkedDaapdUpdater(self.hass, self._api)
```
We should move the updater outside the entity and store the updater in hass.data if needed. Create the `api` instance first and then pass that to both the master entity and the updater when instantiating those.

uvjustin replied:
Use signals to tie updater to master lifecycle?

MartinHjelmare replied:
Yes. :+1:

MartinHjelmare commented on tests/components/forked_daapd/test_media_player.py:
```diff
@@ -0,0 +1,612 @@
+    ) as mock_api:
+        mock_api.return_value.get_request.side_effect = get_request_side_effect
+        mock_api.return_value.add_to_queue.side_effect = add_to_queue_side_effect
+        hass.config_entries._entries.append(config_entry)
```
```suggestion
        config_entry.add_to_hass(hass)
```

MartinHjelmare commented on tests/components/forked_daapd/test_media_player.py:
```diff
@@ -0,0 +1,612 @@
+        return get_request_return_values[update_type]
+
+    async def add_to_queue_side_effect(uris, playback, playback_from_position=None):
+        await master._updater._update(["queue", "player"])  # sorry for touching private
```
Patch `start_websocket_handler` instead. Then we can check the call arguments of that call and get access to the `updater._update` method. Then we can use that method to trigger updates.

MartinHjelmare commented on tests/components/forked_daapd/test_media_player.py:
```diff
@@ -0,0 +1,612 @@
+        await hass.config_entries.async_setup(config_entry.entry_id)
+        await hass.async_block_till_done()
+        master = hass.data[DOMAIN][HASS_DATA_MASTER_KEY]
+        await master._updater._update(["player", "outputs", "queue"])
```
See above.

MartinHjelmare commented on tests/components/forked_daapd/test_media_player.py:
```diff
@@ -0,0 +1,612 @@
+        master = hass.data[DOMAIN][HASS_DATA_MASTER_KEY]
+        await master._updater._update(["player", "outputs", "queue"])
+        await hass.async_block_till_done()
+    return master
```
We should not interact with the entities at all in the tests. We just want to set up the integration and then assert core state and core servies, while patching the library as needed.

MartinHjelmare commented on tests/components/forked_daapd/test_media_player.py:
```diff
@@ -0,0 +1,612 @@
+
+async def test_bunch_of_stuff_master(master, get_request_return_values):
+    """Run bunch of stuff."""
+    await master.async_turn_on()
```
Call the corresponding services instead.

MartinHjelmare commented on tests/components/forked_daapd/test_media_player.py:
```diff
@@ -0,0 +1,612 @@
+    await master.async_media_play()
+
+
+async def test_show_form(hass):
```
Move all the config flow tests to a module `test_config_flow.py`.

uvjustin commented on homeassistant/components/forked_daapd/config_flow.py:
```diff
@@ -0,0 +1,169 @@
+    vol.Optional(CONF_PORT, default=DEFAULT_PORT): int,
+    vol.Optional(CONF_PASSWORD, default=""): str,
+    vol.Optional(CONF_DEFAULT_VOLUME, default=DEFAULT_VOLUME): float,
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): str,
```
Ok, I'll take it out from the form. Is it still ok to pass the zeroconf name through user_data as a default in case it's set up that way?

MartinHjelmare replied:
Why do we need that? We can use the default name. There will only be one instance.

uvjustin replied:
Just thought it was a little prettier for the lazy to have the name that's already set up in forked-daapd prefilled. Not necessary.

uvjustin commented on homeassistant/components/forked_daapd/config_flow.py:
```diff
@@ -0,0 +1,169 @@
+            return self.async_show_form(
+                step_id="user",
+                data_schema=vol.Schema(fill_in_schema_dict(user_input)),
+                errors={"base": validate_result},
```
I understand the need to have an additional decoupler in case the library changes, but for now the decoupler will just be a simple dict that maps the existing strings that the pypi library returns to copies of themselves (since they currently match the strings.json keys). Is that ok?

MartinHjelmare replied:
Yes.

uvjustin commented on homeassistant/components/forked_daapd/manifest.json:
```diff
@@ -0,0 +1,10 @@
+  "domain": "forked_daapd",
+  "name": "forked-daapd",
+  "documentation": "https://www.home-assistant.io/integrations/forked-daapd",
+  "dependencies": [],
```
I assume no need to include aiohttp or asynctest? Is there a list of libraries which we can assume are available?

MartinHjelmare replied:
The requirements for the core and tests will be available. See the requirements*.txt files.

uvjustin commented on homeassistant/components/forked_daapd/media_player.py:
```diff
@@ -0,0 +1,669 @@
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up forked-daapd from a config entry."""
+    name = config_entry.data.get(CONF_NAME)
+    host = config_entry.data.get(CONF_HOST)
```
OK, will try to keep an eye on it. Sometimes hard to keep track of what is there or not.
BTW, since everything in the config has defaults should the vol schema entries be changed from optional to required?

MartinHjelmare replied:
No, optional means that the user doesn't have to fill those items in.

uvjustin commented on homeassistant/components/forked_daapd/media_player.py:
```diff
@@ -0,0 +1,669 @@
+    forked_daapd_master.async_on_remove(
+        async_dispatcher_connect(hass, SIGNAL_ADD_ZONES, async_add_zones)
+    )
+    forked_daapd_master.update_options(config_entry.options)  # configure options
```
Pass on construction and keep in the master entity?

MartinHjelmare replied:
Yes! :+1:

uvjustin commented on homeassistant/components/forked_daapd/media_player.py:
```diff
@@ -0,0 +1,669 @@
+
+async def update_listener(hass, entry):
+    """Handle options update."""
+    hass.data[DOMAIN][HASS_DATA_MASTER_KEY].update_options(entry.options)
```
If the entity has the config_entry, the listener can just be moved inside

MartinHjelmare replied:
It can't be an instance method though due to the use of weak refs in the core.

uvjustin replied:
I see, listener then

uvjustin commented on homeassistant/components/forked_daapd/media_player.py:
```diff
@@ -0,0 +1,669 @@
+        self._available = bool(new_output)
+        if self._available:
+            self._output = new_output
+        self.async_schedule_update_ha_state()
```
Basically if not using force_refresh, we should just use async_write_ha_state instead of async_schedule_update_ha_state?

MartinHjelmare replied:
Yes, exactly, if we're in async context.

uvjustin commented on homeassistant/components/forked_daapd/media_player.py:
```diff
@@ -0,0 +1,669 @@
+                self.hass, SIGNAL_UPDATE_OUTPUTS, self._async_update_output_callback
+            )
+        )
+        self.hass.data[DOMAIN][HASS_DATA_OUTPUTS_KEY].append(self)
```
Don't store the master entity either then?

MartinHjelmare replied:
We don't need to store the master when using the dispatcher to signal, right?

uvjustin committed:
More changes as per code review

Avoid direct access to entities in tests
Bump pypi version
Mark entities unavailable when websocket disconnected
Move config tests to test_config_flow
Move full url creation from media_image_url to library
Move updater entity from master to hass.data
Remove default unmute volume option
Remove name from config_flow
Remove storage of entities in hass.data
Use async_write_ha_state
Use signal to trigger update_options
Use unittest.mock instead of asynctest.mock

MartinHjelmare commented on tests/components/forked_daapd/test_config_flow.py:
```diff
@@ -0,0 +1,147 @@
+            assert result["data"][CONF_PASSWORD] == config_data[CONF_PASSWORD]
+
+            # remove entry
+            await config_entry.async_unload(hass)
```
Make a separate test for each case please. Each combination of entry and exit of the config flow should be one test.

MartinHjelmare replied:
We can remove the unload.

MartinHjelmare commented on tests/components/forked_daapd/test_config_flow.py:
```diff
@@ -0,0 +1,147 @@
+                CONF_TTS_VOLUME: 0.8,
+                CONF_PIPE_CONTROL: "",
+            },
+        )
```
We should assert the result at the end of each test.

MartinHjelmare commented on homeassistant/components/forked_daapd/config_flow.py:
```diff
@@ -0,0 +1,172 @@
+                return self.async_create_entry(
+                    title=f"{FD_NAME} server", data=user_input
+                )
+            validate_result = ERROR_LOOKUP_DICT.get(validate_result, "unknown_error")
```
Let's use the dict inside `validate_input` instead, so the return value of that method is always normalized to home assistant api.

MartinHjelmare commented on homeassistant/components/forked_daapd/media_player.py:
```diff
@@ -0,0 +1,673 @@
+                )
+            )
+        else:
+            _LOGGER.error("Invalid websocket port.")
```
We don't end logging messages with period.
```suggestion
            _LOGGER.error("Invalid websocket port")
```

MartinHjelmare commented on homeassistant/components/forked_daapd/media_player.py:
```diff
@@ -0,0 +1,673 @@
+            _LOGGER.debug(
+                "database/update notifications neither requested nor supported"
+            )
+        await self.hass.async_block_till_done()  # make sure callbacks done before requesting state update
```
Should we instead have the entities signal the updater when they are ready, and use an `asyncio.Event` to have the updater wait for the signal to set the event to allow the updater to continue?

uvjustin replied:
Sure, I can make the change to use asyncio events.

MartinHjelmare commented on homeassistant/components/forked_daapd/media_player.py:
```diff
@@ -0,0 +1,673 @@
+        self.hass = hass
+        self._api = api
+        self.websocket_handler = None
+        self._all_output_ids = []
```
If output ids are unique we should use a set to store them.

uvjustin replied:
Yeah, we can use sets.

MartinHjelmare commented on homeassistant/components/forked_daapd/media_player.py:
```diff
@@ -0,0 +1,673 @@
+        """Private update method."""
+
+        def intersect(list1, list2):  # local helper
+            intersection = [i for i in list1 + list2 if i in list1 and i in list2]
```
Why not make sets of the two lists and use the builtin intersection method for sets?

MartinHjelmare commented on homeassistant/components/forked_daapd/media_player.py:
```diff
@@ -0,0 +1,673 @@
+        if intersect(["outputs", "volume"], update_types):  # update outputs
+            outputs = (await self._api.get_request("outputs"))["outputs"]
+            await self._add_zones(outputs)
+            async_dispatcher_send(self.hass, SIGNAL_UPDATE_OUTPUTS, outputs)
```
Since we already gave the new entities the latest outputs, we shouldn't need to signal those again to update. We can just have them update state based on the output they already have. We only need to signal the old entities to update. This way we don't need to wait until the new entities have been added before we signal the old entities to update.

uvjustin replied:
Sure.
With respect to entity lifecycle, can we assume users cannot delete the entities manually? As I understand it, the integration lifecycle begins with async_setup_entry and ends with the async_unload_entry. Entities only get permanently removed on async_unload_entry then too (but they can get occasional calls to async_added_to_hass and async_will_remove_from_hass, eg when renaming). Is this correct?

MartinHjelmare replied:
Yes, the entity will be removed and re-added if the user changes the entity_id. Also if the user disables an entity, the entity will be removed, or enables a disabled entity, the entity will be added again.

MartinHjelmare commented on tests/components/forked_daapd/test_media_player.py:
```diff
@@ -0,0 +1,555 @@
+        config_entry.add_to_hass(hass)
+        await config_entry.async_setup(hass)
+        await hass.async_block_till_done()
+        print(mock_api.return_value.get_request.call_args)
```
Stale debug print.

MartinHjelmare commented on tests/components/forked_daapd/test_media_player.py:
```diff
@@ -0,0 +1,555 @@
+                ATTR_MEDIA_CONTENT_ID: "somefile.mp3",
+            },
+            blocking=False,
+        )  # don't block and gather task later so hass_async_block_till_done() calls inside _update don't get stuck
```
This is a sign that we shouldn't use `hass.async_block_till_done` in the updater code.

uvjustin replied:
Best practice to avoid using the call generally?

MartinHjelmare replied:
Yeah, we normally only use it in tests.

MartinHjelmare commented on tests/components/forked_daapd/test_media_player.py:
```diff
@@ -0,0 +1,555 @@
+    await updater_update(["player"])
+    # test media play
+    with patch(
+        "homeassistant.components.forked_daapd.media_player.asyncio.sleep",
```
We shouldn't patch sleep. If the code under test uses sleep and we want the tests to be fast, we should patch the time that is slept.

MartinHjelmare commented on tests/components/forked_daapd/test_media_player.py:
```diff
@@ -0,0 +1,555 @@
+    """Test websocket disconnection."""
+    updater_disconnected = mock_api_object.start_websocket_handler.call_args[0][4]
+    updater_disconnected()
+    await hass.async_block_till_done()
```
We need to assert something in each test.

MartinHjelmare commented on tests/components/forked_daapd/test_media_player.py:
```diff
@@ -0,0 +1,555 @@
+            create_task(coro).cancel()
+            raise AsyncioTimeoutError
+
+        with patch("asyncio.wait_for", autospec=True) as wait_for_mock:
```
We should avoid patching asyncio methods. It's too unpredictable since it's standard library and used by the core of home assistant.

MartinHjelmare commented on tests/components/forked_daapd/test_media_player.py:
```diff
@@ -0,0 +1,555 @@
+def test_master_state(hass, mock_api_object):
+    """Test master state attributes."""
+    state = hass.states.get(TEST_MASTER_ENTITY_NAME)
+    print(hass.states.get(TEST_ZONE_ENTITY_NAMES[0]))
```
Stale debug print.

uvjustin committed:
Yet more changes as per code review

Add more assertions in tests
Avoid patching asyncio
Make off state require player state stopped
Only send update to existing zones
Split up some tests
Use events instead of async_block_till_done
Use sets instead of lists where applicable
Wait for pause callback before continuing TTS

MartinHjelmare commented on homeassistant/components/forked_daapd/config_flow.py:
```diff
@@ -0,0 +1,173 @@
+            return self.async_show_form(
+                step_id="user", data_schema=self.discovery_schema, errors={}
+            )
+        await self.async_set_unique_id(CONFIG_FLOW_UNIQUE_ID)  # only allow one entry
```
@balloob is it ok to set a constant unique id of only one entry is allowed?

balloob replied:
It is allowed and we do it for flows that need 1 per network (Sonos, Cast etc). But I don't see a reason in this case why it's limited to a single entry? We have a host so we could connect to multiple ?

balloob replied:
Looking at the screenshot that @escoand posted, it seems like we can get a unique ID from the zeroconf discovery (the E545… part):

![image](https://user-images.githubusercontent.com/1444314/81122047-f017ca00-8ee4-11ea-8243-db06fe05933d.png)

In that case we can use the `zeroconf` step to add unique IDs to existing entries that have been created manually. This is fine since unique ID for config flow is currently only used to be able to ignore discoveries anyway.

See `tradfri` for an example of how this is done:

https://github.com/home-assistant/core/blob/dev/homeassistant/components/tradfri/config_flow.py#L92-L102

uvjustin replied:
Thanks for the feedback.
I originally had it as 1 server per network because in my understanding people only use 1 server, similar to Sonos. That said, there's no reason we can't try to support multiple servers using unique_ids.
However, the _dacp._tcp advertisement posted is not the correct service for detecting a forked-daapd server. There are several services advertised by the server, but the _dacp._tcp one does not seem to include enough information to distinguish a forked-daapd server from other different types of iTunes servers.
The _daap._tcp advertisement which was used does have a Machine ID field which is similar but distinct from the DbId (E545..) field in the _dacp._tcp advertisement. However, looking at the forked-daapd source, both the Machine ID and DbId fields seem to be different hashes (djb hash vs murmur hash) of the library name, so they would not presumably not be unique.
Any suggestions on what else to use for a unique_id?

uvjustin replied:
Since there do not seem to be any true unique ids candidates, it seems we either:
1) Go ahead with the current way which only allows 1 server and just revisit this if anyone requests support for multiple servers.
2) Use the library name or Machine ID as the unique_id and just understand that this is not truly unique and would be problematic if two servers have the exact same library name or if an existing library name gets changed.
Let me know which one you guys choose and I'll put the necessary changes through.

uvjustin commented on tests/components/forked_daapd/test_config_flow.py:
```diff
@@ -0,0 +1,159 @@
+            autospec=True,
+        ) as mock_get_request:
+            mock_get_request.return_value = SAMPLE_CONFIG
+            mock_test_connection.return_value = Future()
```
I will try removing it but I think removing it might give a failure during 3.7 testing. Possibly related to https://bugs.python.org/issue38857

MartinHjelmare replied:
We just want to set the `return_value`to a string, right? That should work.

uvjustin replied:
It works now, thanks. Importing from tests.async_mock does the trick by importing from asynctest for 3.7 and unittest for 3.8.
Made the rest of the changes you mentioned below except for the one with the constant unique_id question.

MartinHjelmare commented on tests/components/forked_daapd/test_config_flow.py:
```diff
@@ -0,0 +1,159 @@
+    with patch(
+        "homeassistant.components.forked_daapd.config_flow.ForkedDaapdAPI.test_connection"
+    ) as mock_test_connection:
+        with patch(
```
It's enough with one `with` statement for both patches. Separate the patch instances with comma.

uvjustin committed:
Remove unnecessary use of Future()

uvjustin commented:
@MartinHjelmare Do you use forked-daapd yourself? I actually don't use it too much myself yet, but it seemed like a good fit for HA because it does a good job of playing/distributing audio in sync to many outputs.
My main use case was piping in audio from a spotify client (I have been using librespot-java as I found it the most up to date) to the forked-daapd server, so that's why I added the option to send basic control commands to the librespot-java pipe source.
I'm thinking it might be good to register each pipe (and maybe the most popular playlists too) as a separate media source. Then users could switch back and forth between different pipes and the normal library this way instead of through config options, and they could get a little more rudimentary control of the what's playing via this selecting of playlists. Further control/queue manipulation would still need to be done at the source itself.
In this case we would need to associate each of the pipe sources with another control api (eg right now the spotify pipe has the librespot-java api), and I'm not sure of the best way to input this association, as the number and naming of pipes would vary across installations (I don't know what other kinds of pipes people use and would be supported). Can we set up the config options to be dynamic with one entry per each pipe?
Curious as to your thoughts on this esp if you happen to use or are familiar with forked-daapd.

MartinHjelmare commented:
I haven't used forked-daapd.

The config options flow can access the integration instances in hass.data. So if there's a controller we could save it in hass.data and query it for what pipes there are.

MartinHjelmare commented:
I suggest we dont make any more changes to this PR unless we have to. I think it's ready and we just need input on my two questions. Further changes can be done in future PRs.

uvjustin commented:
> 
> 
> I suggest we dont make any more changes to this PR unless we have to. I think it's ready and we just need input on my two questions. Further changes can be done in future PRs.

I made the changes already and they didn't involve changing any config options. I can stash them for now.
The result ends up using pipes and playlists as different sources. Is selecting playlists through select source too much of a hack?

uvjustin committed:
Add pipes and playlists as sources

uvjustin committed:
Add support for multiple servers

Change config options to add max_playlists+remove use_pipe_control
Create Machine ID in test_connection and also get from zeroconf
Modify hass.data storage
Update host for known configurations
Use Machine ID in unique_ids, entity names, config title, signals

MartinHjelmare commented:
We shouldn't add unique ids that are not unique. If there's no good unique id, I think we should not allow more than one config entry.

uvjustin commented:
OK, I can roll back the changes.
What about using the config_id as suggested here? https://github.com/home-assistant/architecture/issues/333
We could use the config_id as the basis for the entity/signal names and just use the Machine ID as an identifier to check whether the zeroconf needs to be aborted to avoid unwanted notifications (no harm in non uniqueness resulting in a false positive here).
________________________________
From: Martin Hjelmare <notifications@github.com>
Sent: Tuesday, May 12, 2020, 1:51 AM
To: home-assistant/core
Cc: uvjustin; Mention
Subject: Re: [home-assistant/core] Add forked_daapd integration (#31953)


We shouldn't add unique ids that are not unique. If there's no good unique id, I think we should not allow more than one config entry.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/home-assistant/core/pull/31953#issuecomment-626855491>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AK7SUVL3DB4NR3AVPHMLIV3RRA3LPANCNFSM4KXGLTCA>.

MartinHjelmare commented:
Yes, the config entry.entry_id is ok to use as unique_id for entity unique_id if there's nothing better.

We also need to make sure that the user can't set up the same device again in the config flow.

uvjustin commented:
I got the error "Config flows that are discoverable need to set a unique ID". Can I use a UUID for the config_flow unique_id or should I remove zeroconf support?

MartinHjelmare commented:
IP address may not be used as unique_id as those can change.

uvjustin commented:
Yes, they can change, but if you allow only one config per host then they are unique, no? Is it a requirement they be immutable too?

MartinHjelmare commented:
Yes, the ids may not change.

uvjustin committed:
Use entry_id as basis for multiple entries

uvjustin commented:
Ok, just used the "Machine Name" (library name) for the zeroconf unique id. If there happens to be another instance with the exact same "Machine Name" they won't get a zeroconf flow and can just set up manually.

MartinHjelmare commented on homeassistant/components/forked_daapd/media_player.py:
```diff
@@ -0,0 +1,858 @@
+class ForkedDaapdUpdater:
+    """Manage updates for the forked-daapd device."""
+
+    def __init__(self, hass, api, entity_id):
```
The last parameter should be `entry_id`.

MartinHjelmare commented on homeassistant/components/forked_daapd/const.py:
```diff
@@ -0,0 +1,85 @@
+        "async_media_next_track": "player_next",
+    }
+}
+SIGNAL_ADD_ZONES = "forked-daapd_add_zones"
```
Please add a string formatting argument specifier `{}` to the string so we can use `str.format` to add the entry id to the signal.

MartinHjelmare commented on homeassistant/components/forked_daapd/config_flow.py:
```diff
@@ -0,0 +1,196 @@
+                if same_host_entries:
+                    await asyncio.wait(
+                        [
+                            self.hass.config_entries.async_remove(entry_id)
```
We should not remove existing entries automatically. Show an error, `already_configured`, if the same host is entered.

MartinHjelmare commented on homeassistant/components/forked_daapd/media_player.py:
```diff
@@ -0,0 +1,858 @@
+        if pipes:
+            self._sources_uris.update(
+                {
+                    pipe["title"] + " (pipe)": pipe["uri"]
```
Please use f-strings instead of string concatenation.

MartinHjelmare commented on homeassistant/components/forked_daapd/media_player.py:
```diff
@@ -0,0 +1,858 @@
+        async_add_entities(zone_entities, False)
+
+    remove_add_zones_listener = async_dispatcher_connect(
+        hass, SIGNAL_ADD_ZONES + str(config_entry.entry_id), async_add_zones
```
Please use str.format instead of string concatenation.

uvjustin committed:
Use f-strings and str.format, abort for same host

MartinHjelmare reviewed (approved):
Looks good!

uvjustin committed:
Clean up check for same host

uvjustin commented:
Thanks for all the help @MartinHjelmare

uvjustin commented:
@cgarwood Hi, I have a small commit to clean up some volume setting stuff. Should I submit another PR?

cgarwood commented:
Yeah, submit it as a new PR

nicx commented:
@uvjustin is this media_player supporting TTS? I just tried it but the TTS control is not showing up in the UI. As far as I understand the media player it just needs to support the "SUPPORT_PLAY_MEDIA" flag. I wanted to change my "MDP" Player to this new integration.

uvjustin commented:
> @uvjustin is this media_player supporting TTS? I just tried it but the TTS control is not showing up in the UI. As far as I understand the media player it just needs to support the "SUPPORT_PLAY_MEDIA" flag. I wanted to change my "MDP" Player to this new integration.

The server entity supports play_media but the zone entities do not.
